formats/json/jvmMain/src/kotlinx/serialization/json/JvmStreams.kt

```diff
@@ -0,0 +1,76 @@
+package kotlinx.serialization.json
+
+import kotlinx.serialization.*
+import kotlinx.serialization.json.internal.*
+import java.io.*
+import java.nio.charset.Charset
+
+/**
+ * Serializes the [value] with [serializer] into a [stream] using JSON format and given [charset].
+ *
+ * @throws [SerializationException] if the given value cannot be serialized to JSON.
+ * @throws [IOException] If an I/O error occurs and stream can't be written to.
+ */
+@ExperimentalSerializationApi
+public fun <T> Json.encodeToStream(
+    serializer: SerializationStrategy<T>,
+    value: T,
+    stream: OutputStream,
+    charset: Charset = Charsets.UTF_8
+) {
+    val result = JsonToWriterStringBuilder(stream, charset)
+    try {
+        val encoder = StreamingJsonEncoder(
+            result, this,
+            WriteMode.OBJ,
+            arrayOfNulls(WriteMode.values().size)
+        )
+        encoder.encodeSerializableValue(serializer, value)
+    } finally {
+        result.release()
+    }
+}
+
+/**
+ * Serializes given [value] to [stream] using [charset] and serializer retrieved from the reified type parameter.
+ *
+ * @throws [SerializationException] if the given value cannot be serialized to JSON.
+ * @throws [IOException] If an I/O error occurs and stream can't be written to.
+ */
+@ExperimentalSerializationApi
+public inline fun <reified T> Json.encodeToStream(
+    value: T,
+    stream: OutputStream,
+    charset: Charset = Charsets.UTF_8
+): Unit =
+    encodeToStream(serializersModule.serializer(), value, stream, charset)
+
+/**
+ * Deserializes JSON from [stream] using [charset] to a value of type [T] using [deserializer].
+ *
+ * @throws [SerializationException] if the given JSON input cannot be deserialized to the value of type [T].
+ * @throws [IOException] If an I/O error occurs and stream can't be read from.
+ */
+@ExperimentalSerializationApi
+public fun <T> Json.decodeFromStream(
+    deserializer: DeserializationStrategy<T>,
+    stream: InputStream,
+    charset: Charset = Charsets.UTF_8
+): T {
+    val lexer = JsonReaderLexer(stream, charset)
+    val input = StreamingJsonDecoder(this, WriteMode.OBJ, lexer)
+    val result = input.decodeSerializableValue(deserializer)
+    lexer.expectEof()
```

Comment:
Probably we shouldn't expect eof here since there are use-cases for deserializing more than one object from stream (i.e. produce `Sequence<T>`)

Reply:
Right, this seems to be beyond the scope of our parsing.
In our string-related code, this assertion verified our own parsing logic is correct, but it's not our responsibility for streams.

I've checked out Jackson and it ignores content right after finished object as well

Reply:
Moshi will leave your stream intact allowing you to do things like JSON-RPC which has newline-delimited objects. Granted, we also require that you pass in a buffered stream so that we can read only what's necessary and leave everything else in the buffer you then have access to.

Reply:
I have a use case for implementing a `Flow<T>` from a long-lived HTTP response stream.  To have the lexer expect the end of the stream here seems wrong - why impose this opinion on the stream?  The lexers interest in the stream should begin and end with the fully-formed JSON content IMO

Reply:
having thought about the streaming case I would say that a flow of objects would be a valid case to handle. Probably flow is the right type - both for encoding and decoding. There are some interesting interactions with coroutines there though and flows should probably have their own entrypoint.

Reply:
As a separate matter, these are provided as extension functions. It would be interesting to have an interface like `StringFormat` or `BinaryFormat` (or a combination interface) that provides the consistency between formats. There is also a case to support the `Reader` and `Writer` interfaces next to/instead of InputStream/OutputStream (that would also avoid bringing in charset explicitly while still allowing non-utf8 to be used).